File: src/10.graph/39.126.ts
export function findLadders(
    beginWord: string,
    endWord: string,
    wordList: string[]
): string[][] {
    const wordSet = new Set(wordList);

    if (!wordSet.has(endWord)) {
        return [];
    }

    wordSet.delete(beginWord);

    // 广度优先遍历构建图, key: 单词, value: 在广度优先遍历的第几层
    const steps = new Map<string, number>();
    steps.set(beginWord, 0);

    // 记录单词是由那些单词扩展而来
    const from = new Map<string, Set<string>>();

    // 开始广搜
    let step = 0;
    let found = false;

    const queue: string[] = [beginWord];

    while (queue.length) {
        step++;
        const size = queue.length;
        for (let i = 0; i < size; i++) {
            const curWord = queue.shift();

            for (let j = 0; j < curWord.length; j++) {
                for (let k = "a".charCodeAt(0); k <= "z".charCodeAt(0); k++) {
                    const nextWord =
                        curWord.slice(0, j) +
                        String.fromCharCode(k) +
                        curWord.slice(j + 1);

                    if (steps.has(nextWord) && steps.get(nextWord) === step) {
                        from.get(nextWord).add(curWord);
                    }

                    if (!wordSet.has(nextWord)) {
                        continue;
                    }

                    wordSet.delete(nextWord);

                    queue.push(nextWord);

                    from.set(nextWord, new Set());
                    from.get(nextWord).add(curWord);
                    steps.set(nextWord, step);

                    if (nextWord === endWord) {
                        found = true;
                    }
                }
            }
        }

        if (found) {
            break;
        }
    }

    const ans: string[][] = [];

    if (found) {
        const onPath: string[] = [];
        onPath.push(endWord);
        const dfs = (curWord: string) => {
            if (curWord === beginWord) {
                ans.push([...onPath]);
                return;
            }

            for (const val of from.get(curWord)) {
                onPath.unshift(val);
                dfs(val);
                onPath.shift();
            }
        };

        dfs(endWord);
    }

    return ans;
}

File: src/10.graph/39.126.go
package graph

func FindLadders(beginWord string, endWord string, wordList []string) [][]string {
	ans := [][]string{}
	wordSet := make(map[string]bool)
	for _, v := range wordList {
		wordSet[v] = true
	}
	if !wordSet[endWord] {
		return ans
	}
	wordSet[beginWord] = false

	steps := make(map[string]int)
	steps[beginWord] = 0

	from := make(map[string]map[string]bool)

	step := 0
	found := false

	queue := []string{beginWord}

	for len(queue) > 0 {
		step++
		size := len(queue)
		for i := 0; i < size; i++ {
			curWord := queue[0]
			queue = queue[1:]

			for j := 0; j < len(curWord); j++ {
				for k := 'a'; k <= 'z'; k++ {
					nextWord := curWord[:j] + string(k) + curWord[j+1:]

					if val, ok := steps[nextWord]; ok && val == step {
						from[nextWord][curWord] = true
					}

					if !wordSet[nextWord] {
						continue
					}

					wordSet[nextWord] = false

					queue = append(queue, nextWord)

					from[nextWord] = make(map[string]bool)
					from[nextWord][curWord] = true

					steps[nextWord] = step

					if nextWord == endWord {
						found = true
					}
				}
			}
		}

		if found {
			break
		}
	}

	var dfs func(curWord string)
	onPath := []string{endWord}
	dfs = func(curWord string) {
		if curWord == beginWord {
			ans = append(ans, append([]string{}, onPath...))
			return
		}

		for val := range from[curWord] {
			onPath = append([]string{val}, onPath...)
			dfs(val)
			onPath = onPath[1:]
		}
	}

	if found {
		dfs(endWord)
	}

	return ans
}
